feat: add kubernetes deployment configuration using helm

jcstryker opened the pull request:
This PR resolves #55

Alternative Helm chart to the chart provided here https://github.com/nbcloudio/headplane-chart - I have some concerns around best practices and security - feeling it best to start from scratch to also resolve licensing concerns.

This is provided to this project under the MIT licence.

Thank you to
@antoniolago for doing some fantastic work updating that chart from nbcloudio

@definitelynobody for some further enhancements to the nbcloudio chart - I did not adapt anything directly for this first version of the chart but happy to take a look and help bring some of those features in the future

---

I have tested OIDC Login, Persistence, Admin Login, Database-backed ACLs - I am successfully serving 5-10 devices covering common use cases like exit nodes with this chart in my Homelab.

Notably missing is the tailscale relay feature from the @nbcloudio/@antoniolago/@definitelynobody version of the chart.

I am happy to look into adding that feature in a followup PR once I have time to understand what it does, or someone else can take that enhancement.

---

example `values.yaml` used for testing
```
headplane:
  config:
    url: https://vpn.example.com
    debug: true
    generateCredentials: true
    cookieSecret:
      secretName: vpn-cookie-secret
    oidc:
      enabled: true
      issuerUrl: https://id.example.com
      clientId: foo
      clientSecret:
        secretName: "vpn-oidc-credentials"
      disableApiKeyLogin: true
  persistence:
    enabled: true

headscale:
  config:
    url: https://vpn.example.com
    dns:
      baseDomain: secretnet.example.com
    oidc:
      enabled: true
      issuerUrl: https://id.example.com
      clientId: foo
      clientSecret:
        secretName: "vpn-oidc-credentials"
  persistence:
    enabled: true

extraObjects:
  - apiVersion: gateway.networking.k8s.io/v1
    kind: HTTPRoute
    metadata:
      name: headplane
    spec:
      hostnames:
        - vpn.example.com
      parentRefs:
        - group: gateway.networking.k8s.io
          kind: Gateway
          name: headplane
          namespace: headplane
      rules:
        - backendRefs:
            - group: ""
              kind: Service
              name: headplane
              port: 80
              weight: 1
          matches:
            - path:
                type: PathPrefix
                value: /admin
        - backendRefs:
            - group: ""
              kind: Service
              name: headscale
              port: 80
              weight: 1
          matches:
            - path:
                type: PathPrefix
                value: /
```

---
Follow Up PR ideas:

- Github Action to Publish to OCI repository
- Add Tailscale Relay feature

jcstryker committed:
feat: initial helm chart migration into headplane upstream

Signed-off-by: jcstryker <code@jcstryker.com>

definitelynobody commented:
For reference I added some stuff on top of @antoniolago's repo here

- https://github.com/definitelynobody/headplane-chart

I only was trying to get things working, use my changes however you would like or not at all :)

jcstryker committed:
feat: add support for hostAliases in deployment

Signed-off-by: jcstryker <code@jcstryker.com>

jcstryker committed:
Merge branch 'tale:main' into feat/add-helm-chart

jcstryker committed:
fix: check for public headscale url

Signed-off-by: jcstryker <code@jcstryker.com>

jcstryker committed:
feat: move headscale config to configmap

Signed-off-by: jcstryker <code@jcstryker.com>

jcstryker committed:
feat: add empty split dns map

Signed-off-by: jcstryker <code@jcstryker.com>

jcstryker committed:
chore: enable startup api check by default

Signed-off-by: jcstryker <code@jcstryker.com>

jcstryker committed:
fix(kubernetes): remove unused pvc templates

This commit removes the unused PersistentVolumeClaim templates for headplane and headscale components. The templates were left in place but contained only a TODO comment, indicating they were not yet implemented. Removing them cleans up the codebase and prevents confusion about their intended purpose.

The pvc templates were previously defined in the kubernetes/headplane/templates/headplane/pvc.yaml and kubernetes/headplane/templates/headscale/pvc.yaml files but were not actually needed for the current implementation. This change reduces code clutter and improves maintainability by removing dead code.

Signed-off-by: jcstryker <code@jcstryker.com>

jcstryker committed:
bump versions

Signed-off-by: jcstryker <code@jcstryker.com>

jcstryker committed:
feat: support new secret passing paths

Signed-off-by: jcstryker <code@jcstryker.com>

jcstryker committed:
feat(kubernetes): add persistent volume claims for headplane and headscale

This commit introduces the missing PersistentVolumeClaim templates for both headplane and headscale components. The changes implement the full PVC specification with support for:
- Custom annotations and labels
- Configurable access modes
- Storage requests
- Storage class configuration

The templates follow the existing pattern used in the helm chart and are conditionally enabled based on the persistence configuration values. This provides the necessary storage persistence for both services when enabled.

The implementation includes proper templating with Helm syntax to ensure compatibility with the chart's configuration system.

Signed-off-by: jcstryker <code@jcstryker.com>

jcstryker committed:
fix(kubernetes): correct persistent volume claim name for headscale-data

Signed-off-by: jcstryker <code@jcstryker.com>

tale reviewed (approved):
Looks good to me, the last thing I want to know is how can I deploy this as an OCI native chart so that we can use the `oci://` prefix and also what documentation would look like for using the Helm chart.

tale commented on kubernetes/headplane/Chart.yaml:
```diff
@@ -0,0 +1,6 @@
+description: Kubernetes Helm Chart for Headplane in integrated mode
+name: headplane
+type: application
+version: 1.0.0
```
Can we match it to our SEMver?

tale reviewed (changes requested):
Semver nit

jcstryker committed:
feat: test github action + match semver

Signed-off-by: jcstryker <code@jcstryker.com>

jcstryker committed:
fix: match semver

Signed-off-by: jcstryker <code@jcstryker.com>

jcstryker committed:
feat: allow test branch workflow

Signed-off-by: jcstryker <code@jcstryker.com>

jcstryker committed:
fix: hardcode chart name

Signed-off-by: jcstryker <code@jcstryker.com>

jcstryker committed:
test: chart releaser

Signed-off-by: jcstryker <code@jcstryker.com>

jcstryker committed:
fix: chart releaser

Signed-off-by: jcstryker <code@jcstryker.com>

jcstryker committed:
fix: weird bash error

Signed-off-by: jcstryker <code@jcstryker.com>

jcstryker committed:
chore: move chart publish workflow to separate pr

Signed-off-by: jcstryker <code@jcstryker.com>

tale commented:
I know this might be asking beyond the work you've submitted but do you mind adding links to this in the documentation (`pnpm docs:dev` should spin up a local instance of the website for you to work with), I don't want to merge this in without anyone knowing about it!
